src/shared/util.js

```diff
@@ -1329,9 +1336,221 @@ MessageHandler.prototype = {
+        return pullCapability.promise;
+      },
+      cancel(reason) {
+        let cancelCapability = createPromiseCapability();
+        that.streamControllers[streamId].cancelCall = cancelCapability;
+        that.postMessage({
+          sourceName,
+          targetName,
+          stream: 'cancel',
+          reason,
+          streamId
+        });
+        // Return Promise to signal success or failure.
+        return cancelCapability.promise;
+      }
+    }, queueingStrategy);
+  },
+
+  resolveCall(fn, args, thisArg = null) {
+    if (!fn) {
+      return Promise.resolve(undefined);
+    }
+    return new Promise((resolve, reject) => {
+      resolve(fn.apply(thisArg, args));
+    });
+  },
+
+  _createStreamSink(data) {
+    let that = this;
+    let action = this.actionHandler[data.action];
+    let streamId = data.streamId;
+    let desiredSize = data.desiredSize;
+    let sourceName = this.sourceName;
+    let targetName = data.sourceName;
+    let capability = createPromiseCapability();
+    let sendStreamRequest = ({ stream, chunk, success, reason }) => {
+      this.comObj.postMessage({ sourceName, targetName, stream, streamId,
+                                chunk, success, reason });
+    };
+    let streamSink = {
+      enqueue(chunk, size = 1) {
+        let lastDesiredSize = this.desiredSize;
+        this.desiredSize -= size;
+        // Enqueue decreases the desiredSize property of sink,
+        // so when it changes from positive to negative,
+        // set ready as unresolved promise.
+        if (lastDesiredSize > 0 && this.desiredSize <= 0) {
+          this.sinkCapability = createPromiseCapability();
+          this.ready = this.sinkCapability.promise;
+        }
+        sendStreamRequest({ stream: 'enqueue', chunk });
+      },
+      close() {
+        sendStreamRequest({ stream: 'close' });
+        delete that.streamSinks[streamId];
+      },
+      error(reason) {
+        sendStreamRequest({ stream: 'error', reason });
+      },
+      sinkCapability: capability,
+      onPull: null,
+      onCancel: null,
+      desiredSize,
+      ready: null,
+    };
+    streamSink.sinkCapability.resolve();
+    streamSink.ready = streamSink.sinkCapability.promise;
+    this.streamSinks[streamId] = streamSink;
+    this.resolveCall(action[0], [data.data, streamSink], action[1]).then(() => {
+      sendStreamRequest({ stream: 'start_complete', success: true });
+    }, (reason) => {
+      sendStreamRequest({ stream: 'start_complete', success: false, reason });
+    });
+  },
+
+  _processStreamMessage(data) {
+    let sourceName = this.sourceName;
+    let targetName = data.sourceName;
+    let streamId = data.streamId;
+    let sendStreamResponse = ({ stream, success, reason }) => {
+      this.comObj.postMessage({ sourceName, targetName, stream,
+                                success, streamId, reason });
+    };
+    let resolveOrReject = (capability, success, reason) => {
```

Comment:
You can do `function resolveOrReject(capability, success, reason) {` instead here.